beacon-chain/state/stategen/getter.go

```diff
@@ -109,6 +113,25 @@ func (s *State) StateBySlot(ctx context.Context, slot types.Slot) (*state.Beacon
 	return s.loadStateBySlot(ctx, slot)
 }
 
+// StateByStateRoot retrieves the state with root equal to the input root.
+// It searches for the state only in the last SLOTS_PER_HISTORICAL_ROOT state roots.
+func (s *State) StateByStateRoot(
+	ctx context.Context,
+	stateRoot [32]byte,
+	headState *state.BeaconState,
+) (*state.BeaconState, error) {
+	ctx, span := trace.StartSpan(ctx, "stateGen.StateByStateRoot")
+	defer span.End()
+
+	for i, root := range headState.StateRoots() {
+		if bytes.Equal(root, stateRoot[:]) {
+			slot := ((headState.Slot() / params.BeaconConfig().SlotsPerHistoricalRoot) * params.BeaconConfig().SlotsPerHistoricalRoot).Add(uint64(i))
```

Comment:
you can be more accurate here, how about retrieving the cached block roots stored in the state ? and then using that. Basically look for the cached block root in the same index and you should be able to retrieve the relevant state by `StateByRoot` .

Reply:
Good idea, thanks!